File: source-code/valuations/simple-dividend-discount-model.js
// +------------------------------------------------------------+
//   Model: Simple Dividend Discount Model
//   © Copyright: https://discountingcashflows.com
// +------------------------------------------------------------+

Input(
  {
    _DISCOUNT_RATE: '',
    EXPECTED_DIVIDEND: '',
    _GROWTH_IN_PERPETUITY: '',
    BETA:'',
    _RISK_FREE_RATE: '',
    _MARKET_PREMIUM: '',
    HISTORICAL_YEARS: ''
  },
  [{
    parent:'_DISCOUNT_RATE',
    children:['BETA', '_RISK_FREE_RATE', '_MARKET_PREMIUM']
  }]
); 

$.when(
  get_income_statement(),
  get_income_statement_ltm(),
  get_balance_sheet_statement(),
  get_balance_sheet_statement_quarterly('length:2'),
  get_cash_flow_statement(),
  get_cash_flow_statement_ltm(),
  get_profile(),
  get_dividends_annual(),
  get_prices_annual(),
  get_treasury(),
  get_fx(),
  get_risk_premium()).done(
  function($income, $income_ltm, $balance, $balance_quarterly, $flows, $flows_ltm, $profile, $dividends, $prices, $treasury, $fx, $risk_premium){  
  try{
    var response = new Response({
      income: $income,
      income_ltm: $income_ltm,
      balance: $balance,
      balance_quarterly: $balance_quarterly,
      balance_ltm: 'balance_quarterly:0',
      flows: $flows,
      flows_ltm: $flows_ltm,
      profile: $profile,
      treasury: $treasury,
      dividends: $dividends,
      prices: $prices,
      risk_premium: $risk_premium,
    }).toOneCurrency('income', $fx).merge('_ltm');
    response.balance[0]['date'] = response.prices[0]['date'] = 'LTM';
    
    // +---------------- ASSUMPTIONS SECTION -----------------+ 
    // Count the dividends. If there are no dividends, display a warning.
    var dividendsCount = response.dividends.length - 1;
    if(dividendsCount <= 0){
      throwError("The company does not currently pay dividends!");
      return;
    }
    if(dividendsCount > 10){
    	setAssumption('HISTORICAL_YEARS', 10);
    }
    else{
      // Set the default historical years to the number of historical dividends
      setAssumption('HISTORICAL_YEARS', dividendsCount);
    }
    // Set the growth in perpetuity to the 10 year treasury note
    setAssumption('_GROWTH_IN_PERPETUITY', response.treasury.year10);
    
    // Set beta 
    if(response.profile.beta){
    	setAssumption('BETA', response.profile.beta);
    }
    else{
    	setAssumption('BETA', 1);
    }
    // Risk free rate is the yield of the 10 year treasury note
	setAssumption('_RISK_FREE_RATE', response.treasury.year10);
    setAssumption('_MARKET_PREMIUM', response.risk_premium.totalEquityRiskPremium );
    // Discount Rate is the cost of equity
    setAssumption('_DISCOUNT_RATE', toP(getAssumption('_RISK_FREE_RATE') + getAssumption('BETA') * getAssumption('_MARKET_PREMIUM')));
    
    var sensitivity = 0.01;
    var prefDividendsRatio = absolute((response.income[0].eps * response.income[0].weightedAverageShsOut - response.income[0].netIncome) / response.income[0].netIncome);
    var commonIncome = [];
    if( prefDividendsRatio > sensitivity ){
      commonIncome = ['eps:0', '*', 'weightedAverageShsOut:0'];
    }
    else{
      commonIncome = ['netIncome:0'];
    }
    
    // Setup Original Data
    var original_data = new DateValueData({
      'netIncome': new DateValueList(response.income, 'netIncome'),
      'totalStockholdersEquity': new DateValueList(response.balance, 'totalStockholdersEquity'),
      'weightedAverageShsOut': new DateValueList(response.income, 'weightedAverageShsOut'),
      'eps': new DateValueList(response.income, 'eps'),
      'adjDividend': new DateValueList(response.dividends, 'adjDividend'),
      'marketPrice': new DateValueList(response.prices, 'close'),
    });
    var currentDate = original_data.lastDate();
    var nextYear = currentDate + 1;
    var forecastEndDate = currentDate + 5;
    // chartProjectionYears = 5;
    
    // Compute historical values and ratios
    var historical_computed_data = original_data.setFormula({
      'commonIncome': commonIncome,
      'preferredStockDividends': ['netIncome:0', '-', 'commonIncome:0'],
      'dividendsPaidToCommon': ['adjDividend:0', '*', 'weightedAverageShsOut:0'],
      'bookValue': ['totalStockholdersEquity:0', '/', 'weightedAverageShsOut:0'],
      '_returnOnEquity': ['commonIncome:0', '/', 'totalStockholdersEquity:-1'],
      '_payoutRatio': ['adjDividend:0', '/', 'eps:0'],
      'retainedEarnings': ['eps:0', '-', 'adjDividend:0'],
      '_dividendYield': ['adjDividend:0', '/', 'marketPrice:0'],
      '_adjDividendGrowth': ['function:growth_rate', 'adjDividend'],
      'discountedAdjDividend': ['adjDividend:0'],
    }).compute();
    
    // Compute 5 years of forecasted values and ratios
    var forecasted_data = historical_computed_data.removeDate('LTM').setFormula({
      'linearRegression': ['function:linear_regression', 'adjDividend', {slope: 1, start_date: nextYear - getAssumption('HISTORICAL_YEARS')}],
    }).compute({'forecast_end_date': forecastEndDate});
    
    // By default we want the next year's expected dividend to be the average 
    // between the LTM dividend and the regression value of next year
    var linearRegressionWeight = 0.5;
    var expectedDividend = linearRegressionWeight * forecasted_data.get('linearRegression').valueAtDate(nextYear) + 
        					(1-linearRegressionWeight) * original_data.get('adjDividend').valueAtDate('LTM');
    setAssumption('EXPECTED_DIVIDEND', expectedDividend);
    
    forecasted_data = forecasted_data.setFormula({
      'adjDividend': ['function:compound', getAssumption('EXPECTED_DIVIDEND'), {rate: getAssumption('_GROWTH_IN_PERPETUITY'), start_date: nextYear}],
      '_adjDividendGrowth': ['function:growth_rate', 'adjDividend'],
    }).compute({'forecast_end_date': forecastEndDate});
    // +------------- END OF ASSUMPTIONS SECTION -------------+
    
    // +---------------- MODEL VALUES SECTION ----------------+
    var currency = response.currency;
    // The final value calculated by the Dividend Discount Model
    var valueOfStock = getAssumption('EXPECTED_DIVIDEND') / (getAssumption('_DISCOUNT_RATE') - getAssumption('_GROWTH_IN_PERPETUITY'));
    
    // If we are calculating the value per share for a watch, we can stop right here.
    if(_StopIfWatch(valueOfStock, currency)){
      return;
    }
    _SetEstimatedValue(valueOfStock, currency);
    print(original_data.get('adjDividend').valueAtDate('LTM'), "LTM dividend", '#', currency);
    print(getAssumption('EXPECTED_DIVIDEND'), "Next year's expected dividend", '#', currency);
    print(getAssumption('_DISCOUNT_RATE'), "Cost of Equity", '%');
    print(getAssumption('_GROWTH_IN_PERPETUITY'), "Expected Growth Rate", '%');
    // +------------- END OF MODEL VALUES SECTION ------------+
    
    // +------------------- CHARTS SECTION -------------------+
    forecasted_data.renderChart({
      start_date: nextYear - getAssumption('HISTORICAL_YEARS'),
      keys: ['adjDividend', 'linearRegression'],
      properties: {
        title: 'Historical and Projected Dividends',
        currency: currency,
        disabled_keys: ['linearRegression'],
      }
    });
	// +---------------- END OF CHARTS SECTION ---------------+ 
    
    // +------------------- TABLES SECTION -------------------+
    
    // Dividend Table
    forecasted_data.renderTable({
      start_date: currentDate,
      keys: ['adjDividend', '_adjDividendGrowth'],
      rows: ['{PerShare} Dividends', '{%} Dividend Growth Rate'],
      properties: {
        'title': 'Projected data',
        'currency': currency,
      },
    });
    
    // Historical Table
    historical_computed_data.renderTable({
      start_date: nextYear - getAssumption('HISTORICAL_YEARS'),
      keys: ['netIncome', 'totalStockholdersEquity', '_returnOnEquity', 'dividendsPaidToCommon',
             '_payoutRatio', 'weightedAverageShsOut', 'marketPrice', 'eps', 'adjDividend', '_adjDividendGrowth', '_dividendYield'],
      rows: ['Net income', 'Equity', '{%} Return on equity', 'Dividends paid', 
             '{%} Payout ratio', 'Shares outstanding', '{PerShare} Reference market price', '{PerShare} EPS',
             '{PerShare} Dividends', '{%} Dividend Growth Rate', '{%} Dividend yield'],
      properties: {
        'title': 'Historical data',
        'currency': currency,
        'number_format': 'M',
        'display_averages': true,
        'column_order': 'descending',
      },
    });
    // +---------------- END OF TABLES SECTION ---------------+
  }
  catch (error) {
    throwError(error);
  }
});

Description(`
	<h5>Simple Dividend Discount Model</h5>
	<p>Used to estimate the value of companies that have reached maturity and pay stable dividends as a significant percentage of their Free Cashflow to Equity with little to no high growth chance.</p>
	<p class='text-center'>Read more: <a href='https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#simple-dividend-discount-model-source-code' target='_blank'><i class="fab fa-github"></i> GitHub Documentation</a></p>
`,
  {
    _DISCOUNT_RATE: [
      '{Equation} \\text{Discount Rate} = \\text{Equity Weight} * \\text{Cost of Equity} + \\text{Debt Weight} * \\text{Cost of Debt} * (1 - \\text{Tax Rate})',
      '{Paragraph} Calculated using Weighted Average Cost of Capital (WACC) formula. It represents a firm’s average after-tax cost of capital from all sources, including common stock, preferred stock, bonds, and other forms of debt.',
      '{Link} https://www.investopedia.com/terms/w/wacc.asp',
      
      '{Title} Cost of Equity',
      '{Equation} \\text{Cost of Equity} = \\text{Risk Free Rate} + \\text{Beta} * \\text{Market Premium}',
      '{Paragraph} The cost of equity is the theoretical rate of return that an equity investment should generate. It is calculated using the CAPM formula.',
      '{Link} https://www.investopedia.com/terms/c/costofequity.asp#mntl-sc-block_1-0-20',
      
      '{Title} Cost of Debt',
      '{Equation} \\text{Cost of Debt} = \\frac{\\text{Interest Expense}}{\\text{Total Debt}}',
      '{Paragraph} The cost of debt is the effective rate that a company pays on its debt, such as bonds and loans.',
      '{Link} https://www.investopedia.com/terms/c/costofdebt.asp',
      
      '{Title} Equity & Debt Weights',
      '{Equation} \\text{Debt Weight} = \\frac{\\text{Total Debt}}{\\text{Market Capitalization} + \\text{Total Debt}} = 1 - \\text{Equity Weight}',
      '{Paragraph} The Equity Weight represents the proportion of equity-based financing (Market Capitalization), while the Debt Weight represents the proportion of debt-based financing (Total Debt).',
      
      '{Title} Tax Rate',
      '{Equation} \\text{Tax Rate} = \\frac{\\text{Income Tax Expense}}{\\text{Income Before Tax}}',
      '{Paragraph} The overall tax rate paid by the company on its earned income.',
    ],
    _GROWTH_IN_PERPETUITY: 'The rate at which the company\'s free cash flow is assumed to grow in perpetuity. By default, this is equal to the yield of the U.S. 10 Year Treasury Bond.',
    _OPERATING_CASH_FLOW_MARGIN: [
      '{Equation} \\text{Projected Operating Cash Flow} = \\text{Projected Revenue} * \\text{Operating Cash Flow Margin}',
      'The margin used to project future Operating Cash Flow as a % from future Revenue.',
    ],
    _CAPITAL_EXPENDITURE_MARGIN: [
      '{Equation} \\text{Projected Free Cash Flow} = \\text{Projected Operating Cash Flow} - \\text{Projected Revenue} * \\text{Capital Expedinture Margin}',
      'The margin used to project future Capital Expedinture as a % from future Revenue, which is then used to calculate the Free Cash Flow.',
    ],	
    HISTORICAL_YEARS: 'Number of historical years used to calculate historical averages.',
    REVENUE_REGRESSION_SLOPE: `Future revenues are projected using a linear regression curve of past revenues.
      Set the slope:
      '>1' for a steeper revenue regression curve
      '0' for flat
      '<0' for inverse slope`,
    _RISK_FREE_RATE: 'The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time.'+
    ' By default, it is equal to the current yield of the U.S. 10 Year Treasury Bond.',
    _MARKET_PREMIUM: 'Market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.',
    BETA: 'Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.',
  });

File: source-code/valuations/simple-dividend-discount-model.py
"""
    Model: Simple Dividend Discount Model

    © Copyright discountingcashflows.com
"""

# Render model description
model.render_description(r"""
## Simple Dividend Discount Model

This model estimates the value of companies that have reached maturity and pay stable dividends as a significant percentage of their Free Cash Flow to Equity, with little to no high growth prospects.

See our [GitHub Documentation](https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#simple-dividend-discount-model-source-code)
""")

# Initialize assumptions
assumptions.init({
    "data": {
        "%discount_rate": None,
        "expected_dividend": None,
        "%growth_in_perpetuity": "2.5%",
        "beta": data.get("profile:beta", default=1),
        "%risk_free_rate": data.get("treasury:year10"),
        "%market_premium": data.get("risk:totalEquityRiskPremium"),
        "historical_years": None,
        "forecast_years": 5,
    },
    "hierarchies": [{
        "parent": "%discount_rate",
        "children": ["beta", "%risk_free_rate", "%market_premium"]
    }]
})

# Calculate the discount rate as the cost of equity
risk_free_rate = assumptions.get("%risk_free_rate")
beta = assumptions.get("beta")
market_premium = assumptions.get("%market_premium")
cost_of_equity = risk_free_rate + beta * market_premium

assumptions.set("%discount_rate", cost_of_equity)

# Count all dividend payments excluding None and zero values
dividend_count = data.count("dividend:adjDividend:*", properties={"except_values": [None, 0]})

if dividend_count <= 0:
    raise ValueError("The company does not currently pay dividends!")

elif dividend_count > 10:
    assumptions.set("historical_years", 10)

else:
    # Set the historical years to the number of available dividend payments
    assumptions.set("historical_years", dividend_count)

# Compute the average annual dividend growth rate
data.compute({
    "%dividendGrowth": "function:growth:dividend:adjDividend"
}, forecast=0)

# Define the averaging period for historical dividend growth
averaging_period = f"-{assumptions.get('historical_years')}->0"
average_growth_rate = data.average(f"%dividendGrowth:{averaging_period}")

# Calculate next year's expected dividend by compounding the LTM dividend
# at the average growth rate over the remaining fiscal period (using continuous compounding)
ltm_dividend = data.get("dividend:adjDividend")
if ltm_dividend is not None:
    if average_growth_rate is not None:
        data.compute({
            "dividend:adjDividend": f"""
                function:compound:{ltm_dividend}
                rate:{average_growth_rate}
                continuous:true
            """
        }, forecast=1)
        expected_dividend = data.get(f"dividend:adjDividend:1")
        assumptions.set("expected_dividend", expected_dividend)
    else:
        # Fallback to the ltm_dividend, if the average dividend
        # growth rate could not be computed
        assumptions.set("expected_dividend", ltm_dividend)
else:
    model.error("Zero dividend encountered")

# Calculate the intrinsic stock value using the Dividend Discount Model formula
stock_value = assumptions.get("expected_dividend") / (
    assumptions.get("%discount_rate") - assumptions.get("%growth_in_perpetuity"))

# Set the calculated stock value as the model's final output
model.set_final_value({
    "value": stock_value,
    "units": '$'
})

# Compute additional historical metrics for analysis
data.compute({
    "dividendsPaidToCommon": "dividend:adjDividend * income:weightedAverageShsOut",
    "%returnOnEquity": "income:netIncome / balance:totalStockholdersEquity:-1",
    "%payoutRatio": "dividend:adjDividend / income:eps",
    "%dividendYield": "dividend:adjDividend / quote:close",
    "%adjDividendGrowth": "function:growth:dividend:adjDividend",
})

# Compute projected dividends and dividend growth for the forecast period
data.compute({
    "dividend:adjDividend": f"""
        function:compound:{assumptions.get('expected_dividend')} 
        rate:{assumptions.get('%growth_in_perpetuity')}
        continuous:true
    """,
    "%adjDividendGrowth": "function:growth:dividend:adjDividend",
}, forecast=assumptions.get("forecast_years"))

# Render key results
model.render_results([
    [data.get("dividend:adjDividend:0"), "LTM Dividend", "$"],
    [average_growth_rate, "Average Dividend Growth Rate", "%"],
    [assumptions.get("expected_dividend"), "Next Year's Expected Dividend", "$"],
    [assumptions.get("%discount_rate"), 'Cost of Equity', '%'],
    [assumptions.get("%growth_in_perpetuity"), 'Expected Growth Rate', '%'],
])

# Render a chart showing historical and projected dividends
model.render_chart({
    "data": {
        "dividend:adjDividend": "Dividends",
    },
    "start": -assumptions.get("historical_years"),
    "end": "*",
    "properties": {
        "title": 'Historical and Projected Dividends',
        "include_ltm": True
    }
})

# Render a table for future dividend estimations
model.render_table({
    "data": {
        "dividend:adjDividend": "Dividend per Share",
        "%adjDividendGrowth": "Dividend Growth Rate",
    },
    "start": -1,
    "end": "*",
    "properties": {
        "title": "Future Estimations",
        "order": "ascending",
    },
})

# Render a table of historical financial values and dividend metrics
model.render_table({
    "data": {
        "income:netIncome": "Net Income",
        "balance:totalStockholdersEquity": "Equity",
        "%returnOnEquity": "Return on Equity",
        "dividendsPaidToCommon": "Common Dividends",
        "%payoutRatio": "Payout Ratio",
        "income:weightedAverageShsOut": "Shares Outstanding",
        "quote:close": "Reference Market Price",
        "income:eps": "EPS",
        "dividend:adjDividend": "Dividend per Share",
        "%adjDividendGrowth": "Dividend Growth Rate",
        "%dividendYield": "Dividend Yield",
    },
    "start": -assumptions.get("historical_years"),
    "end": 0,
    "properties": {
        "title": "Historical Values",
        "order": "descending",
        "display_averages": True,
        "number_format": "M"
    },
})

# Set descriptions for assumptions
assumptions.set_description({
    "%discount_rate": r"""
        ## Discount Rate

        `Discount Rate` = `Cost of Equity` = `Risk Free Rate` + `Beta` * `Market Premium`

        The cost of equity is the theoretical rate of return that an equity investment should generate. It is calculated using the CAPM formula.

        [Read More](https://www.investopedia.com/terms/c/costofequity.asp#mntl-sc-block_1-0-20)
    """,

    "beta": r"""
        ## Beta

        Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.
    """,

    "%risk_free_rate": r"""
        ## Risk-Free Rate

        The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time. By default, it is equal to the current yield of the U.S. 10-Year Treasury Bond.
    """,

    "%market_premium": r"""
        ## Market Premium

        The market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.
    """,

    "expected_dividend": r"""
        ## Expected Dividend

        The estimated annual dividend per share for the next year.
    """,

    "%growth_in_perpetuity": r"""
        ## Growth in Perpetuity

        The rate at which the company's free cash flow is assumed to grow indefinitely. By default, this is equal to the yield of the U.S. 10-Year Treasury Bond.
    """,

    "historical_years": r"""
        ## Historical Years

        The number of historical years used to calculate averages for historical data.
    """,

    "forecast_years": r"""
        ## Forecast Years
        
        The number of years into the future for which the analysis projects financial performance and metrics.
    """,
})

print("End of model.")
